File: client/python/examples/python_exec_tty.py
# Run an interactive Python REPL inside the sandbox with a TTY, writing to
# stdin to drive it programmatically.
#
# Run with: python examples/python_exec_tty.py
import asyncio
import sys

import hiver

LINES = [
    "x = 6 * 7",
    "print('the answer is', x)",
    "exit()",
]


async def main() -> None:
    sandbox = await hiver.get_or_create_sandbox(
        "hive-python-exec-tty",
        hiver.SandboxConfig(
            image="hiversh/python:3.13-alpine",
            entrypoint=["tail", "-f", "/dev/null"],
            fs=[
                hiver.LocalFileSystem(
                    backend="local",
                    mount="/workspace",
                    acls=[hiver.ACLRule(path="/workspace/**", access="rw")],
                )
            ],
            ttl=0,
        ),
    )

    exec = await sandbox.exec_stream("python3", cwd="/workspace", tty=True)

    async def feed_stdin() -> None:
        for line in LINES:
            await exec.write_stdin(line + "\r")

    asyncio.create_task(feed_stdin())

    async for pipe in exec.pipes:
        if "stdout" in pipe:
            sys.stdout.write(pipe["stdout"])
        if "stderr" in pipe:
            sys.stderr.write(pipe["stderr"])

    print("\nexit code:", await exec.exit_code)


asyncio.run(main())

File: client/typescript/examples/python-exec-tty.ts
// Run an interactive Python REPL inside the sandbox with a TTY, writing to
// stdin to drive it programmatically.
//
// Run with: npx tsx examples/python-exec-tty.ts
import * as hiver from "@hiver.sh/client";

const sandbox = await hiver.getOrCreateSandbox("hiver-python-exec-tty", {
  image: "python",
  ttl: 0,
});

const exec = await sandbox.execStream("python3", {
  cwd: "/workspace",
  tty: true,
});

// Feed a short script to the REPL line by line, then ask it to exit.
const lines = ["x = 6 * 7", "print('the answer is', x)", "exit()"];

(async () => {
  for (const line of lines) {
    await exec.writeStdin(line + "\r");
  }
})();

for await (const pipe of exec.pipes) {
  if (pipe.stdout) process.stdout.write(pipe.stdout);
  if (pipe.stderr) process.stderr.write(pipe.stderr);
}

console.info("\nexit code:", await exec.exitCode);
